File: FunctionalArithmeticEvaluator/Lexer.fs
﻿module FSharpLexer 

type TokenType =
    | Operation = 0
    | Float     = 1
    
type Token = {
    Type: TokenType
    Value: string
}

[<Literal>]
let Operations = "+-*/%"
[<Literal>]
let Numbers = "1234567890."

let private isOperation (c: char) = Operations.Contains(c)

let private isNumber (c: char) = Numbers.Contains(c)

let rec private getfloatLenght (input:string) lenght =
    match input.Length with
    | 0 -> lenght
    | _ ->
        match input.[0] with
        | c when isNumber c -> getfloatLenght (input.Substring(1)) (lenght + 1)
        | _ -> lenght

let rec tokenize (input:string) =
    match input.Length with
    | 0 -> []
    | _ ->
        match input.[0] with
        | c when isOperation c ->
            let token = { Type = TokenType.Float; Value = input.[0].ToString() }
            token :: (tokenize (input.Substring(1)))
        | c when isNumber c ->
            let floatLenght = getfloatLenght input 0
            let token = { Type = TokenType.Operation; Value = input.Substring(0, floatLenght) }
            token :: (tokenize (input.Substring(floatLenght)))
        | _ -> tokenize (input.Substring(1))

let tailTokenize input : Token list =
    let rec tokenizeRec (input:string) tokens =
        match input.Length with
        | 0 -> []
        | _ ->
            match input.[0] with
            | c when isOperation c ->
                let token = { Type = TokenType.Float; Value = input.[0].ToString() }
                tokenizeRec (input.Substring(1)) (token :: tokens)
            | c when isNumber c ->
                let floatLenght = getfloatLenght input 0
                let token = { Type = TokenType.Operation; Value = input.Substring(0, floatLenght) }
                tokenizeRec (input.Substring(floatLenght)) (token :: tokens)
            | _ -> tokenizeRec (input.Substring(1)) tokens
    tokenizeRec input []

let tailTokenizeChar (input: string) =
    let rec tokenizeRec input visited floatChars =
        let flushFloat = 
            match floatChars with
            | [] -> visited           
            | _ ->
                let floatToken = { Type = TokenType.Float; Value = System.String(floatChars |> List.rev |> Array.ofList) }
                floatToken :: visited
        match input with
        | head :: tail -> 
            match head with
            | c when isOperation c ->            
                let operationToken = { Type = TokenType.Float; Value = head.ToString() }
                tokenizeRec tail (operationToken :: flushFloat) []
            | c when isNumber c -> 
                tokenizeRec tail visited (c :: floatChars)
            | _ -> tokenizeRec tail visited floatChars
        | [] -> List.rev flushFloat
    tokenizeRec (input.ToCharArray() |> List.ofArray) [] []

File: FunctionalArithmeticEvaluator.CSharp/Lexer.cs
using Microsoft.FSharp.Collections;
using System.Diagnostics;

namespace FunctionalArithmeticEvaluator.CSharp
{    
    public static class CSharpLexer
    {
        public enum TokenType
        {
            Operation,
            Float
        }

        [DebuggerDisplay("{Value}")]
        public class Token
        {
            public Token(TokenType type, string value)
            {
                Type = type;
                Value = value;
            }

            public TokenType Type { get; }
            public string Value { get; }
        }

        private const string _operations = "+-*/%";
        private const string _numbers = "1234567890.";
        private static bool IsOperation(this char c) => _operations.Contains(c);
        private static bool IsNumber(this char c) => _numbers.Contains(c);

        private static int GetFloatLenght(this string input, int lenght)
        {
            if (input.Length == 0 || !input[0].IsNumber())
            {
                return lenght;
            }

            return GetFloatLenght(input[1..], lenght + 1);
        }

        public static FSharpList<Token> Tokenize(this string input)
        {
            if (input.Length == 0)
            {
                return ListModule.Empty<Token>();
            }

            var head = input[0];

            if (head.IsOperation())
            {
                var token = new Token(TokenType.Operation, head.ToString());
                return new FSharpList<Token>(token, input[1..].Tokenize());
            }

            if (head.IsNumber())
            {
                var floagLenght = input.GetFloatLenght(0);
                var token = new Token(TokenType.Float, input[0..floagLenght]);
                return new FSharpList<Token>(token, input[floagLenght..].Tokenize());
            }

            return input[1..].Tokenize();
        }
    }
}
